### DISCUSSION TITLE
Initramfs with a long-running service

### DISCUSSION THREAD

**Discussion by rikonor** [dracutdevs/dracut#2624] [Q&A] [answered]
Hi,

I am attempting to create an initramfs that instead of mounting another file-system will have a binary included in it (e.g an HTTP server) that is then started once the initramfs has finished.

This service will then continue running (essentially forever) - I was curious if anyone has done this before, and if so if they can provide any guidance on how to do this.

**Comment by LaszloGombos** [comment]
In the kernel command line 
 - do not specify `root= `
 - specify `init=yourservicebinarypath`
 
Not a common use-case but you might be able to make it work. I have not tried.  Not sure what you mean by "initramfs has finished" as I think you need to basically stay in initramfs.

Some Linux distributions configured to stay in initramfs - e.g. http://tinycorelinux.net

From http://tinycorelinux.net/corebook.pdf
>  it never leaves the first stage. We fully run in the initial RAM disk, never leaving for a spinning disk (or a network mount, etc).

**Reply by rikonor** [reply, marked as answer]
Hi @LaszloGombos - thanks for the info! I think I was overcomplicating things a bit. I probably don't need a tool like dracut for this. What I ended up doing which seems to work fine is:

1. Get a basic Alpine filesystem (I did this by grabbing an official Alpine Docker image and extract the filesystem from it).
1. Inject my binary into this filesystem.
1. Inject an `/init` script that brings up the network and then starts my long-running binary.
1. Create a cpio archive from this filesystem.

I then can start the whole thing with qemu by specifying a kernel and the resulting initramfs.
